src/linker/Linker/WarningSuppressionWriter.cs

```diff
@@ -31,46 +42,87 @@ public void AddWarning (int code, IMemberDefinition memberDefinition)
-						switch (warningOrigin.MetadataToken.TokenType) {
-						case TokenType.TypeDef:
-							sb.Append ("type\", Target = \"");
-							break;
-						case TokenType.Method:
-						case TokenType.Property:
-						case TokenType.Field:
-						case TokenType.Event:
-							sb.Append ("member\", Target = \"");
-							break;
-						default:
-							break;
-						}
-
-						DocumentationSignatureGenerator.Instance.VisitMember (warningOrigin, sb);
-						sb.AppendLine ("\")]");
-					}
-
-					sw.Write (sb.ToString ());
+				if (_fileOutputKind == WarningSuppressionWriterFileOutputKind.Xml)
+					OutputSuppressionsXmlFormat (assemblyName);
+
+				OutputSuppressionsCsFormat (assemblyName);
+			}
+		}
+
+		void OutputSuppressionsXmlFormat (AssemblyNameDefinition assemblyName)
+		{
+			var xmlTree =
+				new XElement ("linker",
+					new XElement ("assembly", new XAttribute ("fullname", assemblyName.FullName)));
+
+			StringBuilder sb = new StringBuilder ();
+			foreach (var warning in GetListOfWarnings (assemblyName)) {
+				DocumentationSignatureGenerator.Instance.VisitMember (warning.Member, sb);
+				xmlTree.Element ("assembly").Add (
+					new XElement ("attribute",
+						new XAttribute ("fullname", "System.Diagnostics.CodeAnalysis.UnconditionalSuppressMessageAttribute"),
+						new XElement ("argument", Constants.ILLink),
+						new XElement ("argument", $"IL{warning.Code}"),
+						new XElement ("argument", GetWarningSuppressionScopeString (warning.Member)),
+						new XElement ("argument", sb.ToString ())));
+
+				sb.Clear ();
+			}
+
+			XDocument xdoc = new XDocument (xmlTree);
+			using (var xw = XmlWriter.Create (Path.Combine (_context.OutputDirectory, $"{assemblyName.Name}.WarningSuppressions.xml"),
+				new XmlWriterSettings { Indent = true })) {
+				xdoc.Save (xw);
+			}
+		}
+
+		void OutputSuppressionsCsFormat (AssemblyNameDefinition assemblyName)
+		{
+			using (var sw = new StreamWriter (Path.Combine (_context.OutputDirectory, $"{assemblyName.Name}.WarningSuppressions.cs"))) {
+				StringBuilder sb = new StringBuilder ("using System.Diagnostics.CodeAnalysis;").AppendLine ().AppendLine ();
+				foreach (var warning in GetListOfWarnings (assemblyName)) {
+					sb.Append ("[assembly: UnconditionalSuppressMessage (\"")
+						.Append (Constants.ILLink)
+						.Append ("\", \"IL").Append (warning.Code)
+						.Append ("\", Scope = \"").Append (GetWarningSuppressionScopeString (warning.Member))
+						.Append ("\", Target = \"");
+
+					DocumentationSignatureGenerator.Instance.VisitMember (warning.Member, sb);
+					sb.AppendLine ("\")]");
 				}
+
+				sw.Write (sb.ToString ());
 			}
 		}
+
+		List<(int Code, IMemberDefinition Member)> GetListOfWarnings (AssemblyNameDefinition assemblyName)
+		{
+			List<(int Code, IMemberDefinition Member)> listOfWarnings = _warnings[assemblyName].ToList ();
+			listOfWarnings.Sort ((a, b) => {
+				string lhs = a.Member is MethodReference lhsMethod ? lhsMethod.GetDisplayName () : a.Member.FullName;
```

Comment:
Will this logic work for members in nested types? I guess Cecil FullName won't compare well with GetDisplayName

Reply:
I think I don't understand why would there be a problem with members in nested types. We only use this strings for sorting the list.

Reply:
I believe GetDisplayName uses `.` as separator whereas FullName uses `+` as separator

Reply:
I think it'd be good to use the signature to sort, since that's what we're outputing.